eth/downloader/queue.go

```diff
@@ -210,96 +310,160 @@ func (q *queue) Insert(headers []*types.Header, from uint64) []*types.Header {
+		// If we've just inserted the fast sync pivot, stop as the following batch needs different insertion
+		if q.mode == FastSync && result.Header.Number.Uint64() == q.fastSyncPivot+1 && len(results) > 0 {
 			break
 		}
-		blocks = append(blocks, block)
-		delete(q.blockPool, block.RawBlock.Hash())
+		results = append(results, result)
+
+		hash := result.Header.Hash()
+		delete(q.blockDonePool, hash)
+		delete(q.receiptDonePool, hash)
 	}
-	// Delete the blocks from the slice and let them be garbage collected
-	// without this slice trick the blocks would stay in memory until nil
-	// would be assigned to q.blocks
-	copy(q.blockCache, q.blockCache[len(blocks):])
-	for k, n := len(q.blockCache)-len(blocks), len(q.blockCache); k < n; k++ {
-		q.blockCache[k] = nil
+	// Delete the results from the slice and let them be garbage collected
+	// without this slice trick the results would stay in memory until nil
+	// would be assigned to them.
+	copy(q.resultCache, q.resultCache[len(results):])
+	for k, n := len(q.resultCache)-len(results), len(q.resultCache); k < n; k++ {
+		q.resultCache[k] = nil
 	}
-	q.blockOffset += uint64(len(blocks))
+	q.resultOffset += uint64(len(results))
 
-	return blocks
+	return results
+}
+
+// ReserveBlocks reserves a set of block hashes for the given peer, skipping any
+// previously failed download.
+func (q *queue) ReserveBlocks(p *peer, count int) *fetchRequest {
+	return q.reserveHashes(p, count, q.hashQueue, nil, q.blockPendPool, len(q.resultCache)-len(q.blockDonePool))
+}
+
+// ReserveNodeData reserves a set of node data hashes for the given peer, skipping
+// any previously failed download.
+func (q *queue) ReserveNodeData(p *peer, count int) *fetchRequest {
+	// Create a task generator to fetch status-fetch tasks if all schedules ones are done
+	generator := func(max int) {
+		q.stateSchedLock.Lock()
+		defer q.stateSchedLock.Unlock()
+
+		if q.stateScheduler != nil {
+			for _, hash := range q.stateScheduler.Missing(max) {
+				q.stateTaskPool[hash] = q.stateTaskIndex
+				q.stateTaskQueue.Push(hash, -float32(q.stateTaskIndex))
+				q.stateTaskIndex++
+			}
+		}
+	}
+	return q.reserveHashes(p, count, q.stateTaskQueue, generator, q.statePendPool, count)
 }
 
-// Reserve61 reserves a set of hashes for the given peer, skipping any previously
-// failed download.
-func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
+// reserveHashes reserves a set of hashes for the given peer, skipping previously
+// failed ones.
+func (q *queue) reserveHashes(p *peer, count int, taskQueue *prque.Prque, taskGen func(int), pendPool map[string]*fetchRequest, maxPending int) *fetchRequest {
 	q.lock.Lock()
 	defer q.lock.Unlock()
 
-	// Short circuit if the pool has been depleted, or if the peer's already
-	// downloading something (sanity check not to corrupt state)
-	if q.hashQueue.Empty() {
-		return nil
-	}
-	if _, ok := q.pendPool[p.id]; ok {
+	// Short circuit if the peer's already downloading something (sanity check not
```

Comment:
Although both should be fine, I'm ok with changing the order to make it a tad simpler form. Will push in the next commit.